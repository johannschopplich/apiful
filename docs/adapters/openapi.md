# `OpenAPI`

The OpenAPI adapter adds type-safety for API calls based on an OpenAPI schema. This includes path names, supported HTTP methods, request body, response body, query parameters, and headers.

In order to use this adapter, `apiverse` needs to generate TypeScript definitions from your OpenAPI schema files beforehand.

> [!NOTE]
> TypeScript definitions with all OpenAPI schema types will be generated by `apiverse`. Make sure to save it as a `.d.ts` file in your project and reference it in your `tsconfig.json` file, if it is not included by default.

## Schema Generation

First, install the [`openapi-typescript`](https://www.npmjs.com/package/openapi-typescript) library:

```bash
# npm
npm install -D openapi-typescript

# pnpm
pnpm add -D openapi-typescript

# yarn
yarn add -D openapi-typescript
```

`apiverse` provides a `generateOpenAPITypes` function that generates a TypeScript definitions file, which you have to save to the proper destination in your project.

Take this `prepare.ts` file as an example:

```ts
import { writeFile } from 'node:fs/promises'
import { defineOpenAPIEndpoints, generateOpenAPITypes } from 'apiverse'

export const endpoints = defineOpenAPIEndpoints({
  petStore: {
    // See: https://petstore3.swagger.io/api/v3/openapi.json
    schema: resolve(currentDir, 'schemas/pet-store.json'),
  },
})

const types = await generateOpenAPITypes(endpoints)
await writeFile('apiverse.d.ts', types)
```

## Using the `OpenAPI` Adapter

After you have generated the TypeScript definitions file, you can use the `OpenAPI` adapter to add type-safety to your API calls:

```ts
import { OpenAPI, createClient } from 'apiverse'

const baseURL = 'https://petstore3.swagger.io/api/v3'
const adapter = OpenAPI<'petStore'>()
const petStore = createClient({ baseURL }).with(adapter)

// The response is typed based on the OpenAPI spec
const userResponse = await petStore('/user/{username}', {
  method: 'GET',
  path: { username: 'user1' },
})
console.log(userResponse)
```

The response returned by the API call on the left is typed as follows:

```ts
declare const userResponse: {
  id?: number
  username?: string
  firstName?: string
  lastName?: string
  email?: string
  password?: string
  phone?: string
  userStatus?: number
}
```

## OpenAPI Path Parameters

OpenAPI can define path parameters on given endpoints. They are typically declared as `/foo/{id}`. Unfortunately, the endpoint type is not defined as `/foo/10`. Thus, using the latter as the path will break type inference.

Instead, use the property `path` to pass an object of the parameters. You can then use the declared path for type inference, and the type checker will ensure you provide all required path parameters. The parameters will be interpolated into the path before the request is made.

```ts
const response = await petStore('/foo/{id}', {
  path: {
    id: 10,
  },
})
```

> [!WARNING]
> Incorrect parameters will not be reported at runtime. An incomplete path will be sent to the backend _as-is_.
