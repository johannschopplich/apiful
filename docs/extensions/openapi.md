# `OpenAPIBuilder`

> [!NOTE]
> This is a [handler extension](/guide/custom-extensions#handler-extension) and wraps [ofetch](https://github.com/unjs/ofetch) under the hood.

This built-in APIful extension adds type-safety for API calls based on an OpenAPI schema. This includes path names, supported HTTP methods, request body, response body, query parameters, and headers.

In order to use this extension, APIful needs to generate TypeScript definitions from your OpenAPI schema files beforehand.

## Prerequisites

To keep the package size small, APIful does not include `openapi-typescript` as a dependency. Install the package using your preferred package manager:

::: code-group
  ```bash [pnpm]
  pnpm add -D openapi-typescript
  ```
  ```bash [yarn]
  yarn add -D openapi-typescript
  ```
  ```bash [npm]
  npm install -D openapi-typescript
  ```
:::

## TypeScript Definitions Generation

Before initiating the type-safe API client, you need to generate TypeScript definitions from your OpenAPI schema files. You can do this by defining API services with OpenAPI schemas in the `apiful.config.ts` file and running the [`generate`](/guide/cli) command of the APIful CLI.

Create an `apiful.config.ts` file and define your API services. For example, let's define a `petStore` service using the OpenAPI schema from the [Swagger Petstore](https://petstore.swagger.io) API:

```ts
import { defineApifulConfig } from 'apiful/config'

export default defineApifulConfig({
  services: {
    petStore: {
      schema: 'https://petstore3.swagger.io/api/v3/openapi.json',
    },
  },
})
```

Then, run the [`generate`](/guide/cli) command in your terminal to generate the TypeScript definitions, saved as `apiful.d.ts`. It augments the `apiful/schema` module with the generated types, so that this extension and you can access the globally defined types.

```sh
npx apiful generate
```

Done! You can now use the `OpenAPIBuilder` extension to create a type-safe API client. Make sure you pass the **service name** to it as a generic parameter, such as `OpenAPIBuilder<'petStore'>()`. Follow the next chapter for more details.

> [!NOTE]
> Make sure the generated `apiful.d.ts` is not excluded by your `tsconfig.json` configuration.

## Using the `OpenAPIBuilder` Extension

After you have generated the TypeScript definitions file, you can use the `OpenAPIBuilder` extension to create a type-safe API client:

```ts
import { createClient, OpenAPIBuilder } from 'apiful'

const baseURL = 'https://petstore3.swagger.io/api/v3'
const adapter = OpenAPIBuilder<'petStore'>()
const petStore = createClient({ baseURL }).with(adapter)
```

Now, the responses are typed based on the OpenAPI specification. For example, the following code snippet fetches a user by username:

```ts
const userResponse = await petStore('/user/{username}', {
  method: 'GET',
  path: { username: 'user1' },
})
```

The response returned by the API call on the left is typed as follows:

```ts
declare const userResponse: {
  id?: number
  username?: string
  firstName?: string
  lastName?: string
  email?: string
  password?: string
  phone?: string
  userStatus?: number
}
```

## OpenAPI Path Parameters

OpenAPI can define path parameters on given endpoints. They are typically declared as `/foo/{id}`. Unfortunately, the endpoint type is not defined as `/foo/10`. Thus, using the latter as the path will break type inference.

Instead, use the property `path` to pass an object of the parameters. You can then use the declared path for type inference, and the type checker will ensure you provide all required path parameters. The parameters will be interpolated into the path before the request is made.

```ts
const response = await petStore('/foo/{id}', {
  path: {
    id: 10,
  },
})
```

> [!WARNING]
> Incorrect parameters will not be reported at runtime. An incomplete path will be sent to the backend _as-is_.

## Request Headers

When working with typed headers in the OpenAPI specification, use the `header` property (not `headers`) to pass your headers:

```ts
const response = await petStore('/some/endpoint', {
  method: 'GET',
  header: {
    'Content-Type': 'application/json',
    'Authorization': 'Bearer <token>',
  }
})
```

The extension will automatically convert this to the correct `headers` property expected by the underlying fetch implementation ([ofetch](https://github.com/unjs/ofetch)).
