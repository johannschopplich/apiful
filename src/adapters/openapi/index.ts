export { OpenAPI } from './client'
export * from '../../openapi/endpoints'
export * from '../../openapi/generate'
export * from '../../openapi/types'
