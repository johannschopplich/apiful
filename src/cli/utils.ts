import type { ApifulConfig } from '../config'
import { loadConfig as _loadConfig } from 'c12'

export function loadConfig(cwd: string) {
  return _loadConfig<Partial<ApifulConfig>>({
    cwd,
    name: 'apiful',
    rcFile: false,
    packageJson: false,
  })
}
