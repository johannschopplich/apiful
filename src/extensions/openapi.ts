import type { OpenAPISchemaRepository } from 'apiful/schema'
import type { ApiClient } from '../client.ts'
import type { OpenAPIClient } from '../openapi/types.ts'
import { ofetch } from 'ofetch'
import { resolvePathParams } from '../openapi/index.ts'

export type * from '../openapi/types.ts'

export type SchemaPaths<K> = K extends keyof OpenAPISchemaRepository
  ? OpenAPISchemaRepository[K]
  : Record<string, never>

export function OpenAPIBuilder<
  const Schema extends string,
  Paths = SchemaPaths<Schema>,
>() {
  return function (client: ApiClient): OpenAPIClient<Paths> {
    const fetcher = ofetch.create(client.defaultOptions)

    return (path, options) => fetcher(
      // @ts-expect-error: Path parameter provided by OpenAPI types
      resolvePathParams(path, options?.path),
      options as Record<string, any>,
    )
  }
}
