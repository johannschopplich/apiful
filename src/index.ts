export * from './adapters'
export * from './client'

export interface OpenAPISchemaRepository {}
