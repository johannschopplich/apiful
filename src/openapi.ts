export { createOpenAPIClient, fetchRequestInterceptor } from './adapters/openapi/client'
