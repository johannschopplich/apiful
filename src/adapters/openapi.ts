import { ofetch } from 'ofetch'
import { resolvePathParams } from '../openapi'
import type { ApiClient } from '../client'
import type { OpenAPIClient } from '../openapi/types'

export interface OpenAPISchemaRepository {}

type ExtractPaths<K> = K extends keyof OpenAPISchemaRepository
  ? OpenAPISchemaRepository[K]
  : Record<string, never>

export function OpenAPI<
  const Schema extends string,
  Paths = ExtractPaths<Schema>,
>() {
  return function (client: ApiClient): OpenAPIClient<Paths> {
    const fetcher = ofetch.create(client.defaultOptions)

    return (path: string, options?: Record<string, any>) =>
      fetcher(
        resolvePathParams(path, options?.path),
        options,
      )
  }
}

export * from '../openapi/types'
