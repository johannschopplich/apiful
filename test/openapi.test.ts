import * as path from 'node:path'
import { describe, expect, it } from 'vitest'
import { defineEndpoints, generateDTS } from '../src/openapi'
import { currentDir } from './utils'

// eslint-disable-next-line test/prefer-lowercase-title
describe('OpenAPI Adapter', () => {
  const endpoints = defineEndpoints({
    sampleApi: {
      schema: path.join(currentDir, 'fixtures/sample-api-schema.yml'),
    },
  })

  it('defines OpenAPI endpoints', () => {
    expect(endpoints).toHaveProperty('sampleApi')
    expect(endpoints.sampleApi).toHaveProperty('schema')
  })

  it('generates OpenAPI types', async () => {
    const types = await generateDTS(endpoints)
    expect(types).toMatchSnapshot()
  })
})
