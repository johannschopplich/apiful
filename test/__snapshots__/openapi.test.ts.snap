// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OpenAPI Adapter > generates OpenAPI types 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck

declare module 'apiful/schema' {
  import { paths as SampleApiPaths } from 'apiful/__sampleApi__'

  interface OpenAPISchemaRepository {
    sampleApi: SampleApiPaths
  }
}

declare module 'apiful/__sampleApi__' {
  export interface paths {
      "/foo/1": {
          parameters: {
              query?: never;
              header?: never;
              path?: never;
              cookie?: never;
          };
          /** Get static foo object with value "1" */
          get: {
              parameters: {
                  query?: never;
                  header?: never;
                  path?: never;
                  cookie?: never;
              };
              requestBody?: never;
              responses: {
                  /** @description A foo object with value "1" */
                  200: {
                      headers: {
                          [name: string]: unknown;
                      };
                      content: {
                          "application/json": {
                              /** @example 1 */
                              foo?: string;
                          };
                      };
                  };
              };
          };
          put?: never;
          post?: never;
          delete?: never;
          options?: never;
          head?: never;
          patch?: never;
          trace?: never;
      };
      "/foo": {
          parameters: {
              query?: never;
              header?: never;
              path?: never;
              cookie?: never;
          };
          /** Get static foo object with value "bar" */
          get: {
              parameters: {
                  query?: never;
                  header?: never;
                  path?: never;
                  cookie?: never;
              };
              requestBody?: never;
              responses: {
                  /** @description A foo object with value "bar" */
                  200: {
                      headers: {
                          [name: string]: unknown;
                      };
                      content: {
                          "application/json": {
                              /** @example bar */
                              foo?: string;
                          };
                      };
                  };
              };
          };
          put?: never;
          post?: never;
          delete?: never;
          options?: never;
          head?: never;
          patch?: never;
          trace?: never;
      };
      "/bar": {
          parameters: {
              query?: never;
              header?: never;
              path?: never;
              cookie?: never;
          };
          get?: never;
          put?: never;
          /** Post request details */
          post: {
              parameters: {
                  query?: never;
                  header?: never;
                  path?: never;
                  cookie?: never;
              };
              requestBody: {
                  content: {
                      "application/json": {
                          [key: string]: unknown;
                      };
                  };
              };
              responses: {
                  /** @description An object containing the request details */
                  200: {
                      headers: {
                          [name: string]: unknown;
                      };
                      content: {
                          "application/json": {
                              /** @example /bar */
                              url?: string;
                              /** @example request body */
                              body?: string;
                              headers?: {
                                  [key: string]: string;
                              };
                              /** @example POST */
                              method?: string;
                          };
                      };
                  };
              };
          };
          delete?: never;
          options?: never;
          head?: never;
          patch?: never;
          trace?: never;
      };
      "/params": {
          parameters: {
              query?: never;
              header?: never;
              path?: never;
              cookie?: never;
          };
          /** Get query parameters */
          get: {
              parameters: {
                  query?: never;
                  header?: never;
                  path?: never;
                  cookie?: never;
              };
              requestBody?: never;
              responses: {
                  /** @description An object containing the query parameters */
                  200: {
                      headers: {
                          [name: string]: unknown;
                      };
                      content: {
                          "application/json": {
                              [key: string]: string;
                          };
                      };
                  };
              };
          };
          put?: never;
          post?: never;
          delete?: never;
          options?: never;
          head?: never;
          patch?: never;
          trace?: never;
      };
  }
  export type webhooks = Record<string, never>;
  export interface components {
      schemas: never;
      responses: never;
      parameters: never;
      requestBodies: never;
      headers: never;
      pathItems: never;
  }
  export type $defs = Record<string, never>;
  export type operations = Record<string, never>;
}
"
`;
